src/Vector/Optional128.hs

```diff
@@ -0,0 +1,152 @@
+{-# language BangPatterns #-}
+{-# language DataKinds #-}
+{-# language DeriveFoldable #-}
+{-# language DeriveFunctor #-}
+{-# language DeriveTraversable #-}
+{-# language DerivingStrategies #-}
+{-# language ExplicitNamespaces #-}
+{-# language GADTSyntax #-}
+{-# language KindSignatures #-}
+{-# language MagicHash #-}
+{-# language RankNTypes #-}
+{-# language ScopedTypeVariables #-}
+{-# language TypeApplications #-}
+{-# language TypeOperators #-}
+{-# language UnboxedTuples #-}
+
+-- This require doing a bunch of unsafe stuff under the hood. The abstraction
+-- offered by natural-arithmetic does not work well for this.
+module Vector.Optional128
+  ( Vector
+  , index
+  , ifoldMap
+  , assign
+  , nothings
+  , update'
+  , map'
+  , ifoldr
+  , itraverse_
+  , traverseST
+  ) where
+
+import Arithmetic.Types (Nat,Fin(Fin),type (<))
+import Control.Monad.ST (ST)
+import Data.Bits (popCount,unsafeShiftL,testBit,countTrailingZeros,clearBit)
+import Data.Bits (setBit)
+import Data.Functor.Classes (Eq1(..))
+import Data.Kind (Type)
+import Data.Primitive (SmallArray)
+import Data.WideWord.Word128 (Word128)
+
+import qualified Arithmetic.Unsafe as Unsafe
+import qualified Data.Primitive as PM
+import qualified Data.Primitive.Contiguous as C
+import qualified GHC.TypeNats as GHC
+
+data Vector :: GHC.Nat -> Type -> Type where
+  Vector :: -- invariant: popcnt(mask) === length(array)
+       !Word128
```

Comment:
Put an `UNPACK` on this field because GHC won't do this for multi-field constructors, even with `-O2`, unless you request it.